### ISSUE TITLE
A really simple example that demos a navigable touchables without any dependencies

### ISSUE THREAD

**Issue by silencer07** [bamlab/react-tv-space-navigation#39]
Hi,

Thank you so much for this library. TV computing has been an underserved platform that deserves love and apps! This framework helps to the realization of that.

I just have a small problem and maybe this is due to me being a new user of this framework. I tried integrating the components in this library and I cannot figure out what is the problem.

Of course I tried reading the example code, and I tried my hardest to mimic what it is doing and I am not successful :( and I don't know why.

I am thinking maybe a small example that is not fancy. Maybe a 3 touchable boxes that is beside one another which you can navigate using tv remote. no outside dependencies, just a bare react native initiated using CLI.

This would help newcomers like me to grasp what this framework requires and how we can make it work within our apps.

Thank you very much and more power to you guys!

P.S. I realized late I am using the standard react native version in my app. do this framework require a forked version like what I see in the example? i.e. **"react-native": "npm:react-native-tvos@0.71.11-0",**

**Comment by pierpo** [comment]
Hey!

About the simple example... that might be a good idea.
We wanted to show a complete example and let you extract what you need.

The thing is: setting up the remote control is not that simple. It needs a few tweaks to be able to catch the native keyevents and plug them to the lib.

The lib's API is kind of simple regarding that, but handling it for multiplatform on the consumer's side isn't that simple.

Can you indicate the problems you encounter? Where do you miss the events?

> P.S. I realized late I am using the standard react native version in my app. do this framework require a forked version like what I see in the example? i.e. "react-native": "npm:react-native-tvos@0.71.11-0",

You can manage without the TV fork because we don't use anything related... except on Apple tvOS 😉 
You'll need to fork to build on this platform, unfortunately.

**Comment by silencer07** [comment]
> Can you indicate the problems you encounter? Where do you miss the events?

Basically my node is not getting even the initial focus even though I properly wrapped it in SpatialNavigationRoot, SpatialNavigationNode and DefaultFocus. I even copy pasted the remote configuration from sample data as well because I thought that is where I am getting it wrong.

TIA!

**Comment by pierpo** [comment]
Oh, this is weird. The fact that you don't even get the initial default focus shows that the issue is not related to the remote!

Do you have any code snippets to share? I'm very curious about this.

About the simplified example, that's kind of what we wanted to do with the tutorial, but it looks like we failed haha

**Comment by silencer07** [comment]
This is the sample code. I truncated it because my code is bad and this is a hobby project I am trying to get deployed before christmas 🗡️ 

`
<SpatialNavigationRoot isActive={isFocused}>
      <SafeAreaView style={styles.container}>
        <View
          style={{
            flexBasis: 1,
            flexGrow: 2,
            flexDirection: "row",
          }}
        >
          <View
            style={{
              flex: 4,
              position: "relative",
              padding: 5,
            }}
          >
            <DefaultFocus>
              <SpatialNavigationNode isFocusable>
                {({ isFocused }) => (
                  <View
                    style={{
                      alignItems: "center",
                      justifyContent: "space-evenly",
                      flexDirection: "row",
                      paddingHorizontal: "25%",
                      width: "100%",
                    }}
                  >
                    <DefaultFocus>
                      <CustomButton
                        style={styles.nextButton}
                        disabled={isEmpty(songs)}
                        onPress={() =>
                          videoPlayerRef.current?.next(currentSongId!)
                        }
                      >
                        <Text
                          style={{
                            color: "lightgray",
                            textAlign: "center",
                            fontSize: 12,
                          }}
                        >
                          Next song
                        </Text>
                      </CustomButton>
                    </DefaultFocus>
                    <CustomButton
                      style={styles.linkButton}
                      onPress={() =>
                        navigation.navigate(RootRoutes.Instructions)
                      }
                    >
                      <Text
                        style={{
                          color: "black",
                          textAlign: "center",
                          fontSize: 12,
                        }}
                      >
                        Link Remote
                      </Text>
                    </CustomButton>
                    <CustomButton
                      style={styles.resetButton}
                      onPress={onResetButtonPress}
                    >
                      <Text
                        style={{
                          color: "white",
                          textAlign: "center",
                          fontSize: 12,
                        }}
                      >
                        Reset
                      </Text>
                    </CustomButton>
                  </View>
                )}
              </SpatialNavigationNode>
            </DefaultFocus>
            .....
`

I copy-pasted the remote configuration and made sure to put it inside the entrypoint of my project
`
import "../components/configureRemoteControl";
`

Thank you very much!

**Comment by manjunath-nuveb** [comment]
`
<SpatialNavigationRoot isActive={isFocused}>
 `
 when are you setting isFocused to true?

**Comment by silencer07** [comment]
@manjunath-nuveb it is from reacat navigation `const isFocused = useIsFocused();`

Just assume it is always true for the sake of code snippet :D

**Comment by pierpo** [comment]
@manjunath-nuveb  that's a good check indeed 😊  thanks!

@silencer07 I don't really understand what could be wrong here 🤔 

I would remove the TouchableOpacity from the buttons. It might mess up with the lib because the whole lib relies on ignoring native focus 😄 

Also, can you try a simpler example in your codebase and check whether it works or not?
First, I'd make a simpler button : 

```
const Button = ({ onSelect, label }) => (
 <SpatialNavigationNode
      onSelect={onSelect}
      orientation="horizontal"
      isFocusable={true}
    >
      {({ isFocused }) => {
        return (
          <Text style={{ backgroundColor: isFocused ? 'white' : 'black' }}>{label}</Text>
        );
      }}
    </SpatialNavigationNode>
)

```

Then, I'd use it in the most simple page:

```tsx
<SpatialNavigationRoot>
  <SpatialNavigationNode>
    <Button
      onPress={() => console.log('A')}
      label="A"
    />
    <Button
      onPress={() => console.log('B')}
      label="B"
    />
  </SpatialNavigationNode>
</SpatialNavigationRoot>
```

**Comment by silencer07** [comment]
Thanks @pierpo. I would definitely try this on friday and get back to you. I decided to take a two day break for now :D

I really love how you are so engage in this!

**Comment by silencer07** [comment]
HI @pierpo

I did install `react-native-keyevent` ! Else the code will be complaining about missing dependencies.

Yes I did copy all the variants of RemoteControlManager which include the android version. see the screenshot below

<img width="326" alt="Screenshot 2023-12-12 at 6 13 56 AM" src="https://github.com/bamlab/react-tv-space-navigation/assets/19424606/4c9ed398-7374-4924-8f19-f779147a85b9">

**Comment by silencer07** [comment]
Oh my! I was able to make it work. You are right the `react-native-keyevent` needs integrated native code.

since I am using expo, I just needed to use https://github.com/ChronSyn/react-native-keyevent-expo-config-plugin and configure it!

This is lovely! I think I am unstuck now!

P.S. your remote configuration is a sensible default, I am thinking maybe we should integrate it to the framework itself.

**Comment by silencer07** [comment]
Alright. After fixing the code, my next problem is "onFocus" event not working

This is my custom button now

`const highlightedStyle = {
  borderWidth: 2,
  borderColor: "lightgray",
  borderStyle: "solid",
};

export function CustomButton(props: TouchableOpacityProps) {
  return !Platform.isTV ? (
    <TouchableOpacity {...props}>{props.children}</TouchableOpacity>
  ) : (
    <SpatialNavigationNode
      onFocus={props.onFocus as () => void}
      onBlur={props.onBlur as () => void}
      onSelect={props.disabled ? (props.onPress as () => void) : undefined}
      isFocusable={true}
    >
      {({ isFocused }) => {
        let style: Record<string, any> = props.style || {};
        if (isFocused) {
          style = { ...style, ...highlightedStyle };
        }
        return <View style={style}>{props.children}</View>;
      }}
    </SpatialNavigationNode>
  );
}`

and this is how I use it

`<CustomButton
                        style={styles.linkButton}
                        onPress={() =>
                          navigation.navigate(RootRoutes.Instructions)
                        }
                      >
                        <Text
                          style={{
                            color: "black",
                            textAlign: "center",
                            fontSize: 12,
                          }}
                        >
                          Link Remote
                        </Text>
                      </CustomButton>`
                      
 weirdly enough nothing happens when you press the enter on remote

**Comment by pierpo** [comment]
Oh, this is probably related to react-native-keyevent indeed. I can't help with that. You should add breakpoints in the native code to see if the enter key is handled?

**Comment by silencer07** [comment]
Gotcha, I will just file a new issue in the said library and get back to you. Let's leave this issue open until we found a solution @pierpo

**Comment by silencer07** [comment]
@pierpo 

I did some more testing, and it seems that doing a long press on center button of remote actually register an "enter" key event. 

while this unblocks my further development, It would be nice if there is a workaround I can do to register short press as long press for the mean time :)

**Comment by silencer07** [comment]
@pierpo  someone pointed me on this one: https://github.com/kevinejohn/react-native-keyevent/issues/52

I will play around and see if this works. if yes, I think we will need a `patch-package`d version of react-native-keyevent

**Comment by ionictest2017** [comment]
> ```tsx
> <SpatialNavigationRoot>
>   <SpatialNavigationNode>
>     <Button
>       onPress={() => console.log('A')}
>       label="A"
>     />
>     <Button
>       onPress={() => console.log('B')}
>       label="B"
>     />
>   </SpatialNavigationNode>
> </SpatialNavigationRoot>
> ```

i have try this simple in example project but it is not working in android while focus is working with TouchableOpacity

**Comment by silencer07** [comment]
> @pierpo someone pointed me on this one: [kevinejohn/react-native-keyevent#52](https://github.com/kevinejohn/react-native-keyevent/issues/52)
> 
> ticket that i filed in react-native-keyevent library: [kevinejohn/react-native-keyevent#80](https://github.com/kevinejohn/react-native-keyevent/issues/80)
> 
> I will play around and see if this works. if yes, I think we will need a `patch-package`d version of react-native-keyevent

@pierpo seems that there is an additional code needed in the android native side for DPAD_CENTER of a remote to work

https://github.com/ChronSyn/react-native-keyevent-expo-config-plugin/issues/4

I am thinking maybe you can update your example?

**Comment by pierpo** [comment]
@silencer07 sorry I had not answered. Thank you for your work on this 😊 
We're keeping the issue open and we'll add this to the example indeed.

We just switched the example to Expo so it's convenient to have your patch 😁 
(And we added a patch already to support Expo 50)

**Comment by manjunath-nuveb** [comment]
Additional code needed in  https://github.com/kevinejohn/react-native-keyevent

https://github.com/kevinejohn/react-native-keyevent/issues/52#issuecomment-688061593

**Comment by manjunath-nuveb** [comment]
My issue is similar to this. but happens only when virtual keyboard is opened and closed
https://github.com/kevinejohn/react-native-keyevent/issues/83

**Comment by 0xHexE** [comment]
```
diff --git a/build/withReactNativeKeyevent.js b/build/withReactNativeKeyevent.js
index 8ba81fda107fdcff0beb2d2c924c25874b988fe9..8f605f6bcf7df8037bfd23bf9e1fbc50aeddb786 100644
--- a/build/withReactNativeKeyevent.js
+++ b/build/withReactNativeKeyevent.js
@@ -99,6 +99,15 @@ const withAndroidMainActivityBody = (config) => {
     // @ts-ignore
     const newConfig = (0, config_plugins_1.withMainActivity)(config, (config) => {
         const newSrc = [
+            '@Override',
+            'override fun dispatchKeyEvent(event: KeyEvent): Boolean {',
+            '    if ((event.keyCode == KeyEvent.KEYCODE_ENTER || event.keyCode == KeyEvent.KEYCODE_DPAD_CENTER) && event.action == KeyEvent.ACTION_DOWN) {',
+            '        KeyEventModule.getInstance().onKeyDownEvent(event.keyCode, event)',
+            '        return false',
+            '    }',
+            '    return super.dispatchKeyEvent(event)',
+            '}',
+
             'override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {',
             '   // // Uncomment this if key events should only trigger once when key is held down',
             '  // if (event.getRepeatCount() == 0) {',

```

Patch for the expo 50 in order this to work with @bam.tech/react-native-keyevent-expo-config-plugin
